pysrc/bytewax/connectors/kafka/registry.py

```diff
@@ -68,74 +73,50 @@ def _get_schema(self, schema_ref: Union[int, SchemaRef]) -> str:
-            instance.
-
-        """
-        self._base_url = base_url
-
-    def _get_schema(self, schema_ref: Union[int, SchemaRef]) -> bytes:
-        # Schema can bew retrieved by `schema_id`, or by
-        # specifying a `subject` and a `version`, which
-        # defaults to `latest`.
-        if isinstance(schema_ref, int):
-            url = f"{self._base_url}/schemas/{schema_ref}/schema"
+        if serde_format == "confluent-avro":
+            return ConfluentAvroSerializer(self.client, schema)
+        elif serde_format == "plain-avro":
+            return PlainAvroSerializer(schema)
         else:
-            version = schema_ref.version or "latest"
-            url = (
-                f"{self._base_url}/subjects/"
-                f"{schema_ref.subject}/versions/"
-                f"{version}/schema"
-            )
-        resp = requests.get(url)
-        resp.raise_for_status()
-        return resp.content
-
-    def serializer(
-        self, schema_ref: Union[int, SchemaRef]
-    ) -> SchemaSerializer[Dict, bytes]:
-        """Fastavro serializer.
-
-        :arg schema_ref: Specify either the `schema_id` or a
-            `SchemaRef` instance.
-
-        """
-        schema = self._get_schema(schema_ref)
-        return _AvroSerializer(schema)
+            msg = f"Unsupported serde_format: {serde_format}"
+            raise ValueError(msg)
 
     def deserializer(
-        self, schema_ref: Union[int, SchemaRef]
+        self,
+        schema_ref: Optional[Union[int, SchemaRef]] = None,
+        serde_format: SerdeFormats = "confluent-avro",
     ) -> SchemaDeserializer[MaybeStrBytes, AvroMessage]:
-        """Fastavro deserializer.
+        """Confluent avro deserializer.
 
-        :arg schema_ref: Specify either the `schema_id` or a
-            `SchemaRef` instance.
+        `schema_ref` is optional if you choose the default "confluent-avro"
+        serde_format, since Confluent's serialization library adds the schema_id as
+        metadata in each message. The client will automatically fetch and cache the
+        schemas needed.
 
+        If you chose "plain-avro", you instead need to define a single schema that will
+        be used for each message.
         """
-        schema = self._get_schema(schema_ref)
-        return _AvroDeserializer(schema)
+        if serde_format == "confluent-avro":
+            if schema_ref is not None:
+                _logger.warning("schema_ref supplied to deserialized will be ignored")
+            return ConfluentAvroDeserializer(self.client)
+        elif serde_format == "plain-avro":
+            assert (
+                schema_ref is not None
+            ), "schema_ref required for 'plain-avro' deserialization format"
+            schema = self._get_schema(schema_ref)
+            return PlainAvroDeserializer(schema)
+        else:
+            msg = f"Unsupported serde_format: {serde_format}"
+            raise ValueError(msg)
```

Comment:
> Schema Logic - Where do I get the schema I want to use for a given message? Is it magic bytes in the Kafka message? Does the registry know it because of the topic name? Is it a manual convention and it needs to be manually specified? I'm sure you could come up with other conventions too.

I think that's the reason why confluent's adds the magic bytes. You usually need to know the schema_id beforehand, or through other, unspecified means. The schema object returned from the registry then tells you its format. So we could instantiate the correct deserializer looking at the schema. The problem is that confluent's schema registry calls their own special format `AVRO`, so we have no way of knowing if a message is plain avro or not from the response of the registry, hence why I added the option at the deserializer instantiation.

I'll look more into docs and api to see if I missed something here

Reply:
Adding some more context, [here](https://www.confluent.io/blog/kafka-connect-deep-dive-converters-serialization-explained/):
> Problem: Reading non-Avro data with AvroConverter
This may be the most common error that I see reported again and again on places like the Confluent Community mailing list and Slack group. It happens when you try to use the Avro converter to read data from a *topic that is not Avro*. **This would include data written by an Avro serializer other than the Confluent Schema Registry’s Avro serializer**, which has its own wire format.

I like how they define "data written by any avro serializer that's not ours" as "not avro". I understand the need for the wire format, but complaining that people want to deserialize correct avro messages with something called `AvroConverter` is funny.

This is the description of the wire format in their docs:
https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#wire-format

The same problem will apply to protobuf if we add support for it.

So, to sum up, the schema is effectively `Avro`, that's why the call it that in the registry, but the messages you serialize with their library are instead in the wire format.

My initial idea was to add the format in `SchemaRef`, so that when you instantiate a (de)serializer you already have all the info you need. The problem with doing that using confluent's library, is that the deserializer only accepts the wire format, and when it sees a different schema_id it will just download the new schema with no way of customizing the behavior. If we instead use only an avro deserializer, it won't work with messages serialized by confluent's library.